Test/Controllers/RoomListViewController.swift

```diff
@@ -31,95 +31,83 @@ class RoomListViewController: UITableViewController {
+//                    return
+//            }
+//
+//            messageViewController.room = room
+//            messageViewController.chatkit = self.chatkit
+//        }
     }
     
     override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return self.roomProvider?.numberOfRooms ?? 0
+        return self.viewModel?.rooms.count ?? 0
     }
 
     override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
         let cell = tableView.dequeueReusableCell(withIdentifier: "roomCell", for: indexPath)
         
-        if let roomCell = cell as? TestTableViewCell {
-            let room = self.roomProvider?.room(at: indexPath.row)
+        if let roomCell = cell as? TestTableViewCell, let viewModel = self.viewModel {
+            let room = viewModel.rooms[indexPath.row]
             
-            roomCell.testLabel.text = room?.name
+            roomCell.testLabel.text = room.name
         }
         
         return cell
     }
 }
 
-extension RoomListViewController: JoinedRoomsProviderDelegate {
+extension RoomListViewController: JoinedRoomsViewModelDelegate {
     
-    func joinedRoomsProvider(_ joinedRoomsProvider: JoinedRoomsProvider, didJoinRoomsAtIndexRange range: Range<Int>) {
-        if self.tableView.numberOfRows(inSection: 0) == 0 {
-            self.tableView.reloadData()
-        }
-        else {
-            self.tableView.beginUpdates()
-            
-            range.forEach {
-                let indexPath = IndexPath(row: $0, section: 0)
-                self.tableView.insertRows(at: [indexPath], with: .fade)
-            }
-            
-            self.tableView.endUpdates()
-        }
+    func joinedRoomsViewModel(_ joinedRoomsViewModel: JoinedRoomsViewModel, didAddRoomAtIndex index: Int, changeReason: JoinedRoomsViewModel.ChangeReason) {
+        self.tableView.beginUpdates()
+        
+        let indexPath = IndexPath(row: index, section: 0)
+        self.tableView.insertRows(at: [indexPath], with: .fade)
+        
+        self.tableView.endUpdates()
     }
     
-    func joinedRoomsProvider(_ joinedRoomsProvider: JoinedRoomsProvider, didUpdateRoomAtIndex index: Int, previousValue: Room) {
+    func joinedRoomsViewModel(_ joinedRoomsViewModel: JoinedRoomsViewModel, didUpdateRoomAtIndex index: Int, changeReason: JoinedRoomsViewModel.ChangeReason) {
         self.tableView.beginUpdates()
         
         let indexPath = IndexPath(row: index, section: 0)
-        self.tableView.reloadRows(at: [indexPath], with: .fade)
+        
+        switch changeReason {
+        case .dataUpdated:
+            self.tableView.reloadRows(at: [indexPath], with: .fade)
+            
+        case let .messageReceived(previousIndex):
+            let previousIndexPath = IndexPath(row: previousIndex, section: 0)
+            self.tableView.moveRow(at: previousIndexPath, to: indexPath)
```

Comment:
Does moving a row automatically reload the data for it too?

Reply:
Yes, it does reload the content of the cell.